Implement profile API

andy-armstrong opened the pull request:
TNL-1491

andy-armstrong commented:
@cahrens Here's my first cut at a profile API. I'm going to move on to the profile page now, but I'd love your feedback when you have time.

cahrens commented on lms/envs/common.py:
```diff
@@ -2050,3 +2050,25 @@
+
+# Default visibility level for accounts without a specified value
+# The value is one of: 'global', 'edx_users', 'private'
+DEFAULT_PROFILE_VISIBILITY = 'edx_users'
```
I don't think we should use "edx". And don't include "global" since we aren't implementing that yet.
I think "all_users", "private".

cahrens commented on lms/envs/common.py:
```diff
@@ -2050,3 +2050,25 @@
+
+######################## Leaner profile configuration ###########################
+
+# Default visibility level for accounts without a specified value
```
I was thinking this setting would be more for a single profile configuration (because we know we may want to support multiple in the future), what is the default visibility, what are the fields that are in the profile, and which fields are always visible. I'd do it as a dict instead of 3 different settings.

cahrens commented on openedx/core/djangoapps/user_api/profiles/__init__.py:
```diff
@@ -0,0 +1,2 @@
+PROFILE_VISIBILITY_PREF_KEY = 'profile_privacy'
```
Perhaps define the different levels of visibility here too?

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,68 @@
+        response = self.send_get(self.client)
+        data = response.data
+        self.assertEqual(2, len(data))
+        self.assertEqual(self.user.username, data["username"])
```
Why are you checking just these 2? The user should be able to see all her fields.

But you should also have a test that DIFFERENT users can only see the 2 always visible fields.

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,68 @@
+        data = response.data
+        self.assertEqual(6, len(data))
+        self.assertEqual(self.user.username, data["username"])
+        self.assertIsNone(data["profile_image"])
```
Set some data on the profile to make this a more interesting test...

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,64 @@
+    """
+        **Use Cases**
+
+            Get the user's public profile information.
```
I find the "public profile" confusing. If the user is calling it, he can get all his information. If it is a different user, it will return only the public information (if user has made rest of profile private).

cahrens commented on lms/envs/common.py:
```diff
@@ -2050,3 +2050,25 @@
+DEFAULT_PROFILE_VISIBILITY = 'edx_users'
+
+# The list of fields that are shown on a shared learner profile
+SHARED_PROFILE_PUBLIC_FIELDS = [
```
I find the "PUBLIC" here to be confusing. I would think of these as just the profile fields.

cahrens commented on lms/envs/common.py:
```diff
@@ -2050,3 +2050,25 @@
+]
+
+# The list of fields that are shown on a private learner profile
+PRIVATE_PROFILE_PUBLIC_FIELDS = [
```
Again private/public is confusing. These are the fields that are always shown in the profile, regardless of the selected visibility setting.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,64 @@
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (permissions.IsAuthenticated,)
+
+    DEFAULT_PUBLIC_PROFILE_FIELD_NAMES = [
```
Why is this here?

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,64 @@
+            public_field_names = settings.PRIVATE_PROFILE_PUBLIC_FIELDS
+        for field_name in public_field_names:
+            profile_settings[field_name] = account_settings.get(field_name, None)
+        return Response(profile_settings)
```
I don't think this is correct-- if the requesting user is same as username, all fields should be returned.

andy-armstrong commented:
Thanks for the great feedback, @cahrens. I've tried to address it all, so please re-review.

cahrens commented on lms/envs/common.py:
```diff
@@ -2050,3 +2050,27 @@
+    # The list of all fields that can be shown on a learner's profile
+    "all_fields": [
+        'username',
+        'profile_image',
```
Some of these names may change (profile_image, languages), but that isn't a big deal.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -29,7 +28,54 @@ def setUp(self):
+        legacy_profile.mailing_address = "Park Ave"
+        legacy_profile.save()
+
+
```
Extra newlines.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -29,7 +28,54 @@ def setUp(self):
+
+
+@ddt.ddt
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
```
Remove the skip now that you put stuff in cms urls?

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -45,6 +91,19 @@ def test_get_account_different_user(self):
         self.different_client.login(username=self.different_user.username, password=TEST_PASSWORD)
         self.send_get(self.different_client, expected_status=404)
 
+    def test_get_account_anonymous_user(self):
```
Why are these methods duplicated?

andy-armstrong replied:
I accidentally moved them into the superclass, then moved them back, and obviously screwed things up in the process. Good catch!

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -82,11 +82,19 @@ def get(self, request, username):
+    @staticmethod
+    def get_serialized_account(username):
+        """
+        Returns the serialized JSON for the specified username.
```
Add in the documentation that callers must do authentication/permissions check before calling this method.

cahrens commented on openedx/core/djangoapps/user_api/legacy_urls.py:
```diff
@@ -0,0 +1,41 @@
+from . import views as user_api_views
+from .models import UserPreference
+
+
```
Extra newlines

andy-armstrong replied:
I'm confused about the newline rule. I thought that two lines were needed after the imports. I guess it is instead that two lines are needed before each class.

andy-armstrong replied:
Interestingly, this file just moved unchanged, so I don't know why GitHub doesn't see that.

cahrens commented on openedx/core/djangoapps/user_api/profiles/__init__.py:
```diff
@@ -0,0 +1,4 @@
+PROFILE_VISIBILITY_PREF_KEY = 'profile_privacy'
+
+ALL_USERS_VISIBILITY = 'all_users'
```
Add doc strings about what these are (how used).

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,89 @@
+from openedx.core.djangoapps.user_api.models import UserPreference
+from openedx.core.djangoapps.user_api.profiles import PROFILE_VISIBILITY_PREF_KEY
+
+TEST_PASSWORD = "test"
```
Perhaps move this into a variable in UserAPITestCase.

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,89 @@
+
+
+@ddt.ddt
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
```
Remove skip now?

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Test that a logged in user can get her own public profile information.
+        """
+        self.create_mock_profile(self.user)
```
Combine this wiht previous test as a ddt test case (see my test for example).

And I still think it would be better to have some info in the profile to make it a more interesting test.

Note that the language property is currently "language", not "languages" (and it exists in the account API). You might as well use that and we can change it later if needed.

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,89 @@
+        self.client.login(username=self.user.username, password=TEST_PASSWORD)
+
+        # Verify that a user with a private profile only returns the public fields
+        UserPreference.set_preference(self.user, PROFILE_VISIBILITY_PREF_KEY, 'private')
```
Also need a test with "different" user when profile visibility is set to private.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,81 @@
+
+            GET /api/user/v0/profiles/{username}/
+
+        **Response Values for GET**
```
Document the query parameter.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,81 @@
+        """
+        GET /api/user/v0/profiles/{username}/[?include_all={true | false}]
+
+        Note: include_all can only be specified if the user is making the request
```
include_all will only be honored if the user...

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,81 @@
+    @staticmethod
+    def get_serialized_profile(username, configuration, include_all_results=False):
+        """
+        Returns a JSON representation of the user's profile settings.
```
Add in comments about checking permissions/authentication before calling this method.

andy-armstrong replied:
I did originally have a default arg for configuration but it caused a circularity that I couldn't resolve. I   realize now that I should have None as the default value, and then or in the true default.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,81 @@
+        profile_settings = {}
+        privacy_setting = ProfileView._get_user_profile_privacy(username, configuration)
+        if include_all_results or privacy_setting == ALL_USERS_VISIBILITY:
+            public_field_names = configuration.get('all_fields')
```
name variable "profile_field_names"?

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,81 @@
+            public_field_names = configuration.get('public_fields')
+        for field_name in public_field_names:
+            profile_settings[field_name] = account_settings.get(field_name, None)
+        return profile_settings
```
Calling this "profile_setitngs" seems a bit odd. Maybe just profile?

andy-armstrong replied:
I had it that way to mirror account_settings, but I agree that it is odd. I've switched it to profile.

andy-armstrong commented:
@cahrens Here's the next round of fixes. I'm going to work on removing the old profile API now.

andy-armstrong commented:
@cahrens Added handling and tests for requests for unknown usernames.

andy-armstrong commented:
@cahrens I've merged your PR and rebased it into this one. I've also remove the profile_info API and fixed the code to use the account API instead.

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -15,12 +15,11 @@
 from student.views import do_email_change_request
-from openedx.core.djangoapps.user_api.accounts import NAME_MIN_LENGTH
 from openedx.core.djangoapps.user_api.accounts.serializers import AccountLegacyProfileSerializer, AccountUserSerializer
+from openedx.core.djangoapps.user_api.api.account import AccountUserNotFound
```
This import already exists below (line 25).

cahrens commented:
The unit test failures mystify me.

andy-armstrong commented:
@cahrens There is some kind of circular import which was happening during start up of the LMS and causing the whole thing to break. Delaying the load of these bad imports until just when they are required resolved the issue. As I said in the TODO, we really shouldn't be importing openedx from common, so maybe there's a cleaner way to fix this logic.

It is interesting that having so many imports in some of these central classes means that almost every class gets loaded at once. I would think that this would be somewhat of a performance issue, although I suppose class loading is mostly very quick. It is counterintuitive that all these classes are loaded even during paver commands though.

andy-armstrong commented:
I don't understand those unit test failures either. I'll try to find time to investigate this weekend.

andy-armstrong commented:
@cahrens I've worked it out. The problem is that the tests only work in LMS, but I had removed the skiptest for the whole suite. I've added it back just to the email test since that needs templates found only in the LMS. Hopefully I'll get a green build now unless I've broken something else in the process.

BTW, it shows how badly interdependent these libraries are that code in common depends upon LMS templates. At some point it would be good to try to clean all this logic up so that the LMS injects the behavior into the platform, rather than having common assume that it is available.

cahrens commented:
Yeah, I thought of the CMS issue as I was falling asleep Friday night. I would prefer we remove the URLs from CMS until we need them and we have a solution for the templates.

cahrens commented:
And you haven't responded to my earlier feedback yet-- look back over the comments (unless you did it after I looked Friday night).

andy-armstrong commented:
@cahrens I don't know how I missed your last set of comments. I've addressed them now. 

I feel conflicted about removing the CMS urls since that is where we want it to go, and it almost all works. I'm going to leave it in for now as I don't have time tonight to remove it and get all the tests sorted out. We can talk about it on Monday.

andy-armstrong commented:
Jenkins, test this please

andy-armstrong commented:
Jenkins, test this please

andy-armstrong commented:
Jenkins, test this please

andy-armstrong commented:
Jenkins, test this please

andy-armstrong commented:
Jenkins, test this please

andy-armstrong commented:
@cahrens I have a green build with the APIs removed from Studio. I've also submitted a change to improve the doc strings. Please review one more time and if it is good to go I'll squash and wait for one more build.

cahrens commented on cms/envs/common.py:
```diff
@@ -915,3 +915,26 @@
 }
+
+# The configuration for Studio user profiles
+PROFILE_CONFIGURATION = {
```
@andy-armstrong I thought you were going to just import this variable from lms' common.py file.

andy-armstrong replied:
This is no longer needed, so I should have removed it. I've done so now.

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -89,14 +87,34 @@ def get(self, request, username):
+
+        Note:
+          This method does not perform authentication so it is up to the caller
+          to ensure that the result is only returned to the appropriate user.
```
I'd still like you to describe what the conditions are for someone being an "appropriate user".

andy-armstrong replied:
It seems like overkill to me to have to document all the rules here. The intention is that the caller has done the authorization and determined that is is appropriate to return the account. There are also callers that don't expose the result so don't have to perform any authorization. I'd rather have the authorization scheme documented in the code that does it.

andy-armstrong replied:
For the record, we decided that it is worth being very explicit here as this is essentially a programmatic version of the API.

cahrens commented on openedx/core/djangoapps/user_api/api/profile.py:
```diff
@@ -151,22 +101,19 @@ def update_email_opt_in(username, org, optin):
+    account_settings = AccountView.get_serialized_account(username)
+    year_of_birth = account_settings['year_of_birth']
     of_age = (
-        profile.year_of_birth is None or  # If year of birth is not set, we assume user is of age.
```
I think we will want to change this assumption (year of birth not set, assume user of age). We certainly plan to take the opposite approach for profile pictures. But that is beyond the scope of this PR....

cahrens replied:
I created https://openedx.atlassian.net/browse/ECOM-1154

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,113 @@
+from openedx.core.djangoapps.user_api.accounts.tests.test_views import UserAPITestCase
+from openedx.core.djangoapps.user_api.models import UserPreference
+from openedx.core.djangoapps.user_api.profiles import PROFILE_VISIBILITY_PREF_KEY
+
```
I'd still rather see TEST_PASSWORD in UserAPITestCase.

andy-armstrong replied:
Sorry, I missed that comment before. It isn't even used, so I've removed it.

cahrens commented on openedx/core/djangoapps/user_api/profiles/tests/test_views.py:
```diff
@@ -0,0 +1,113 @@
+        client = self.login_client(api_client, requesting_username)
+
+        # Verify that a user with a private profile only returns the public fields
+        UserPreference.set_preference(self.user, PROFILE_VISIBILITY_PREF_KEY, 'private')
```
Use constant defined in profiles/__init__.py.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,113 @@
+
+        **Response Values for GET**
+
+            Returns the same responses as for the AccountView API, but
```
It only returns a subset of the fields that the AccountView API returns.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,113 @@
+            filtered based upon the user's specified privacy permissions.
+            If the parameter include_all is passed as 'true' then a user
+            can receive all fields for their own account, ignoring
+            the field visibility preference.
```
I'd suggest adding that the include_all URL parameter is ignored if the profile requested is not for the user making the request.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,113 @@
+        then all profile fields are returned.
+
+        Note:
+          This method does not perform authentication so it is up to the caller
```
Again, I would suggest defining what the "appropriate user" means.

cahrens commented on openedx/core/djangoapps/user_api/profiles/views.py:
```diff
@@ -0,0 +1,113 @@
+
+        Args:
+          username (str): The username for the desired account.
+          configuration (dict):
```
Add a description of the configuration format.

cahrens commented:
:+1: with just a few comments above. You can address the comments and squash all at the same time.
